source/extensions/filters/network/postgres_proxy/postgres_decoder.h

```diff
@@ -0,0 +1,135 @@
+  virtual void incTransactionsCommit() PURE;
+  virtual void incTransactionsRollback() PURE;
+
+  enum class NoticeType { Warning, Notice, Debug, Info, Log, Unknown };
+  virtual void incNotices(NoticeType) PURE;
+
+  enum class ErrorType { Error, Fatal, Panic, Unknown };
+  virtual void incErrors(ErrorType) PURE;
+};
+
+// Postgres message decoder.
+class Decoder {
+public:
+  virtual ~Decoder() = default;
+
+  virtual bool onData(Buffer::Instance& data, bool frontend) PURE;
+  virtual PostgresSession& getSession() PURE;
+};
+
+using DecoderPtr = std::unique_ptr<Decoder>;
+
+class DecoderImpl : public Decoder, Logger::Loggable<Logger::Id::filter> {
+public:
+  DecoderImpl(DecoderCallbacks* callbacks) : callbacks_(callbacks) { initialize(); }
+
+  bool onData(Buffer::Instance& data, bool frontend) override;
+  PostgresSession& getSession() override { return session_; }
+
+  void setMessage(std::string message) { message_ = message; }
+  std::string getMessage() { return message_; }
+
+  void setMessageLength(uint32_t message_len) { message_len_ = message_len; }
+  uint32_t getMessageLength() { return message_len_; }
+
+  void setStartup(bool startup) { startup_ = startup; }
+  void initialize();
+
+  bool encrypted() const { return encrypted_; }
+
+protected:
+  bool parseMessage(Buffer::Instance& data);
+  void decode(Buffer::Instance& data);
+  void decodeAuthentication();
+  void decodeBackendStatements();
+  void decodeBackendErrorResponse();
+  void decodeBackendNoticeResponse();
+  void decodeFrontendTerminate();
+
+  void incMessagesUnknown() { callbacks_->incMessagesUnknown(); }
+
+  void incSessionsEncrypted() { callbacks_->incSessionsEncrypted(); }
+  void incSessionsUnencrypted() { callbacks_->incSessionsUnencrypted(); }
+
+  DecoderCallbacks* callbacks_;
+  PostgresSession session_;
+
+  // the following fields store result of message parsing
+  char command_;
+  std::string message_;
+  uint32_t message_len_;
+
+  bool startup_{true};    // startup stage does not have 1st byte command
+  bool encrypted_{false}; // tells is exchange is encrypted
+
+  // Message action defines the Decoder's method which will be invoked
+  // when a specific message has been decoded.
+  using MsgAction = std::function<void(DecoderImpl*)>;
+
+  // Message has two fields:
+  // field 0 - string with message description
+  // field 1 - vector of Decoder's methods which are invoked when the message
+  // is processed.
+  using Message = std::tuple<std::string, std::vector<MsgAction>>;
+
+  // Frontend and Backend messages
+  // Class could be used to group those values, but tuple is used until
+  // functionality requires a switch.
+  // field 0 - string describing direction (Frontend or Backend)
+  // field 1 - hash map indexed by messages'1 1st byte points to data used for processing messages
+  // field 2 - data used for processing messages not found in hash map
+  std::tuple<std::string, absl::flat_hash_map<char, Message>, Message> FE_messages_;
```

Comment:
I think that these are okay to define struct. Why did you use tuple in this?

Reply:
No real reason. Structure can be used as well. It just binds 3 things together without introducing new types or names. When I was working on it the concept seemed simple enough to put it into one nameless "bag". Does it make reading the code harder? If so, I can refactor it to structs, np.

Reply:
Moved it to _MsgGroup_ struct.